modelskill/observation.py

```diff
@@ -303,124 +338,115 @@ class TrackObservation(Observation):
         quantity: Optional[Quantity] = None,
     ):
-        self._filename = None
-        self._item = None
-
-        if isinstance(data, pd.DataFrame):
-            df = data
-            df_items = df.columns.to_list()
-            items = self._parse_track_items(df_items, x_item, y_item, item)
-            df = df.iloc[:, items].copy()
-        elif isinstance(data, str):
-            assert os.path.exists(data)
-            self._filename = data
-            if name is None:
-                name = os.path.basename(data).split(".")[0]
-
-            ext = os.path.splitext(data)[-1]
-            if ext == ".dfs0":
-                dfs = mikeio.open(data)
-                file_items = [i.name for i in dfs.items]
-                items = self._parse_track_items(file_items, x_item, y_item, item)
-                df, iteminfo = self._read_dfs0(dfs, items)
-                if quantity is None:
-                    quantity = Quantity.from_mikeio_iteminfo(iteminfo)
-            else:
-                raise NotImplementedError(
-                    "Only dfs0 files and DataFrames are supported"
-                )
+        assert isinstance(
+            data, get_args(TrackType)
+        ), "Could not construct TrackObservation from provided data type."
+
+        # self._filename = None
+        # self._item = None
+
+        if isinstance(data, (str, Path)):
+            assert (
+                Path(data).suffix == ".dfs0"
+            ), "File must be a dfs0 file, other file types must be read with pandas or xarray"
+            name = name or Path(data).stem
+            data = mikeio.read(data)  # now mikeio.Dataset
+        elif isinstance(data, mikeio.Dfs0):
+            data = data.read()  # now mikeio.Dataset
+
+        # parse items and convert to xr.Dataset
+        items = [x_item, y_item, item]
+        if isinstance(data, mikeio.Dataset):
+            ds, item_names = self._mikeio_to_xarray(data, items)
+            if quantity is None:
+                iteminfo = data[item_names[2]].item  # 2=value item
+                quantity = Quantity.from_mikeio_iteminfo(iteminfo)
+        elif isinstance(data, pd.DataFrame):
+            ds, item_names = self._pandas_to_xarray(data, items)
+        elif isinstance(data, xr.Dataset):
+            ds, item_names = self._xarray_to_xarrray(data, items)
         else:
             raise TypeError(
-                f"input must be str or pandas DataFrame! Given input has type {type(data)}"
+                f"input must be str, Path, mikeio.Dataset, pd.DataFrame or xr.Dataset! Given input has type {type(data)}"
             )
 
+        name = name or item_names[2]
+        name = self._validate_name(name)
+
+        # make sure that x and y are named x and y
+        old_xy_names = list(ds.data_vars)[:2]
+        ds = ds.rename(dict(zip(old_xy_names, ["x", "y"])))
+
+        vars = [v for v in ds.data_vars if v != "x" and v != "y" and v != "z"]
+        ds = ds.rename({vars[0]: name})
```

Comment:
Why vars[0]?

Is there only a single variable?

Variables in a dataset are not ordered.